### PR TITLE
bf-lu perf updates

### PR DESCRIPTION
**Pull request by vishwacsena**
Made bunch of fixes to improve perf for core bf-lu operations. 

Notable example include large .lu files - e.g. email skill.lu
This multi-file convert now takes ~1,715ms .vs. 5,704ms, better by about 70%
> time bin/run luis:convert --in emailskills.lu
real   0m1.715s           [old: 0m5.704s]
user    0m2.056s .     [old: 0m6.057s]
sys     0m0.179s         [old: 0m0.182s]

The BF-LU tests now take ~89s .vs. ~129s, better by about 30%. 
> time npm run test 
real    1m28.603s     [old: 2m8.290s]
user    0m48.068s   [old: 1m38.726s]
sys     0m2.266s       [old: 0m3.117s]

### CODE REVIEW

File: packages/lu/src/parser/lufile/parseFileContents.js [hunk 1]
```diff
@@ -1538,15 +1551,15 @@ const handleClosedList = function (parsedContent, entityName, listLines, entityR
...
                     throw (new exception(retCode.errorCode.SYNONYMS_NOT_A_LIST, error.toString(), [error]));
                 }
-                if (!nvExists.list.includes(item)) nvExists.list.push(item);
+                nvExists.list.push(item);
             })
         }
     });
-
+    // de-dupe the list
+    if (nvExists && nvExists.list)nvExists.list = [...new Set(nvExists.list)];
```
Comment (feich-ms) [inline packages/lu/src/parser/lufile/parseFileContents.js]: whitespace before nvExists.list = [...new Set(nvExists.list)]

File: packages/lu/src/parser/luis/luis.js [hunk 2]
```diff
@@ -62,25 +63,29 @@ class Luis {
...
         }
-        checkAndUpdateVersion(this)
+        helpers.checkAndUpdateVersion(this)
     }
 }
 
 module.exports = Luis
 
 const initialize = function(instance, LuisJSON) {
+    instance.utteranceHash = {};
```
Comment (munozemilio) [inline packages/lu/src/parser/luis/luis.js]: Why adding this to the Luis object?

File: packages/lu/src/parser/luis/luis.js [hunk 3]
```diff
@@ -104,6 +109,36 @@ const compareString = function(a, b) {
     return 0;
 }
 
+const mergeResultsWithHash = function (blob, finalCollection, type, hashProperty) {
+    if (blob[type].length === 0) { 
+        return
+    }
+    if (!finalCollection.utteranceHash) finalCollection.utteranceHash = {};
```
Comment (munozemilio) [inline packages/lu/src/parser/luis/luis.js]: Why not making this local?

File: packages/lu/src/parser/lufile/parseFileContents.js (hunk 1 above)
Comment (vishwacsena) [reply packages/lu/src/parser/lufile/parseFileContents.js]: Fixed

File: packages/lu/src/parser/luis/luis.js (hunk 3 above)
Comment (vishwacsena) [reply packages/lu/src/parser/luis/luis.js]: Updated this to be local.

**Comment by vishwacsena** [comment]
@munozemilio updated. I also noticed that separating the qna from qna alterations causes all files to parsed twice which is expensive. We should probably have qnabuilder do both qnaList as well as qnaAlterations with a single parse.

File: packages/lu/src/parser/luis/luis.js [hunk 4]
```diff
@@ -58,35 +59,38 @@ class Luis {
...
     }
 }
 
 module.exports = Luis
 
-const initialize = function(instance, LuisJSON) {
+const initialize = function(instance, LuisJSON, hashTable = undefined) {
     for (let prop in LuisJSON) {
         instance[prop] = LuisJSON[prop];
+        if (hashTable !== undefined && (prop === LUISObjNameEnum.UTTERANCE || prop === LUISObjNameEnum.PATTERNS)) {
```
Comment (munozemilio) [inline packages/lu/src/parser/luis/luis.js]: This operation should be done for both scenarios, if Luis object is empty or already initialized.

Comment (vishwacsena) [reply packages/lu/src/parser/luis/luis.js]: Fixed

**Pull request merged**